src/java.base/share/classes/java/lang/StringUTF16.java

```diff
@@ -1486,7 +1481,13 @@ public static void putCharSB(byte[] val, int index, int c) {
 
     public static void putCharsSB(byte[] val, int index, char[] ca, int off, int end) {
         checkBoundsBeginEnd(index, index + end - off, val);
-        putChars(val, index, ca, off, end);
+        String.checkBoundsBeginEnd(off, end, ca.length);
+        Unsafe.getUnsafe().copyMemory(
+                ca,
+                Unsafe.ARRAY_CHAR_BASE_OFFSET + ((long) off << 1),
+                val,
+                Unsafe.ARRAY_BYTE_BASE_OFFSET + ((long) index << 1),
+                (long) (end - off) << 1);
```

Comment:
The documentation of `copyMemory()` is not super-clear about endianness.
But it seems to imply that in this case it behaves as if it were to copy `short`s, so endianness seems to be preserved.

The invocation of `copyMemory()` here implicitly assumes that `ARRAY_CHAR_INDEX_SCALE` and `ARRAY_BYTE_INDEX_SCALE` are 2 and 1, resp., which seems quite reasonable but not written in the stone.

Reply:
I've found the code that imposes this requirement:

https://github.com/openjdk/jdk/blob/7bf4c608e7b23398ea89cf0a592e9554af98144e/src/hotspot/share/opto/library_call.cpp#L1713-L1718